### ISSUE TITLE
Is this expected? CPU utilization goes up and down again and again (from 80% to 7%)

### ISSUE THREAD

**Issue by jasonbigl** [rqlite/rqlite#1393]
Hi,

I'm using rqlite as sqlite replication solution in our cluster, I'm trying to understand why our cpu utilization has this pattern (line goes up and down, from ~7% to ~80% again and again)
![cpu](https://github.com/rqlite/rqlite/assets/2292217/34fce5f3-c5dc-4c0a-928d-c8e00b76c1a8)

The rqlite server is on a pure server that no other apps running, on google cloud platform, machine type is e2-small with 2 vCPUs and 2G of memory, the database size is about ~300Mb, when I check the rqlite logs, I found when cpu spikes, the rqlite is doing something like following:
```
[store] 2023/09/11 09:29:41 node snapshot created in 1.020558246s
2023-09-11T09:29:41.283+0800 [INFO]  raft: starting snapshot up to: index=34069618
2023-09-11T09:29:41.283+0800 [INFO]  snapshot: creating new snapshot: path=/data/raft/snapshots/2-34069618-1694395781283.tmp
2023-09-11T09:29:56.818+0800 [INFO]  snapshot: reaping snapshot: path=/data/raft/snapshots/2-34060768-1694395263321
[store] 2023/09/11 09:29:56 snapshot and persist took 16.565476114s
2023-09-11T09:29:56.826+0800 [INFO]  raft: compacting logs: from=34053166 to=34059413
2023-09-11T09:29:56.833+0800 [INFO]  raft: snapshot complete up to: index=34069618
[store] 2023/09/11 09:31:46 node snapshot created in 966.117457ms
```

So the question is: is this a expected behavior? that rqlite cpu usage goes up and down when syncing data.

Thank you.

**Comment by otoolep** [comment]
What version are you running?

The series 7 release (and earlier) periodically reads the entire SQLite database into memory, compresses it (gzip), and then writes the compressed data to disk. That's probably what you're seeing, assuming you're running 7.x. It does this as part of truncating the Raft log. See the Raft paper to learn why this is done. If you plot disk IO, you'll probably that spike too. 

Version 8 -- which is what is being developed right now on the `master` branch -- is mostly about making this whole process much more efficient. So you can wait for 8.0, or you might like to bump the VM size if you want to throw more CPU at it.

**Comment by jasonbigl** [comment]
thank you very much for the fast reply! I'm using latest version v7.21.4

"If you plot disk IO, you'll probably that spike too" - yes, the disk IO spike too, just like cpu.

ok, before V8 comes out, I will scale the vm size, thank you!

**Comment by otoolep** [comment]
OK, great, if you pay attention to this repo, you'll find out when 8.0 is ready for initial (non-prod) testing. I'm quite close to that point -- hopefully a month away. You could compare performance then and see if it improves.

**Comment by otoolep** [comment]
Actually I just committed an important bug fix to master, and ran load testing myself. master is not production ready yet, but you might be interested in building it, and trying it out. I would very much appreciate other folks banging on it.

https://rqlite.io/docs/install-rqlite/building-from-source/

**Comment by jasonbigl** [comment]
no problem, I would love to test, just want to confirm, will it be ok if I add master as another voter (or non-voter maybe) into our current cluster? our current cluster is on production and on different version v7.21.4

**Comment by jasonbigl** [comment]
another question please, what algorithm does rqlite for snapshot compressing?

in our cluster, the rqlite consume more cpu than the disk IO, when compressing snapshot and logs, the cpu is about to arrive the 100%, while disk IO is only about 400 iops and 5MiB/s, far less than the disk performance limitation (we have 3000 ios and 125MiB/s for our SSD in google cloud platform).

Is there anyway to make this more IO bound not CPU bound? I found this post: https://github.com/hashicorp/consul/issues/9620, maybe can we add a option to set the compressing algorithm?

I'm also tuning the `-compression-batch` option to smaller number to see if it can reduce the cpu load.

Thank you!

**Comment by otoolep** [comment]
>no problem, I would love to test, just want to confirm, will it be ok if I add master as another voter (or non-voter maybe) into our current cluster?

8.0 is not production ready, you risk data loss if you use it in any way in a production cluster. I'm suggesting you test a completely separate 8.0 system.

>another question please, what algorithm does rqlite for snapshot compressing?
gzip, as mentioned above.

>Is there anyway to make this more IO bound not CPU bound? 

You are stuck with the current performance in 7x, as no changes are planned for that series. However one of the main goals for 8.0 is more efficient snapshotting.

>I'm also tuning the -compression-batch option to smaller number to see if it can reduce the cpu load.

That parameter is unrelated to snapshotting so don't bother with it.

**Comment by otoolep** [comment]
Fixed on master, 8.0 will be fixed very soon.

**Comment by otoolep** [comment]
8.x series has been released and should address this.
